table/block_based/block_based_table_builder.cc

```diff
@@ -1501,15 +1501,16 @@ Status BlockBasedTableBuilder::InsertBlockInCache(const Slice& block_contents,
     const size_t read_amp_bytes_per_bit =
         rep_->table_options.read_amp_bytes_per_bit;
 
-    TBlocklike* block_holder = BlocklikeTraits<TBlocklike>::Create(
-        std::move(results), read_amp_bytes_per_bit,
-        rep_->ioptions.statistics.get(),
-        false /*rep_->blocks_definitely_zstd_compressed*/,
-        rep_->table_options.filter_policy.get());
+    std::unique_ptr<TBlocklike> block_holder(
+        BlocklikeTraits<TBlocklike>::Create(
+            std::move(results), read_amp_bytes_per_bit,
+            rep_->ioptions.statistics.get(),
+            false /*rep_->blocks_definitely_zstd_compressed*/,
+            rep_->table_options.filter_policy.get()));
 
     if (block_holder->own_bytes()) {
       size_t charge = block_holder->ApproximateMemoryUsage();
-      s = block_cache->Insert(key, block_holder, charge,
+      s = block_cache->Insert(key, block_holder.release(), charge,
                               &DeleteEntryCached<TBlocklike>);
```

Comment:
Also, it looks like this feature was developed alongside block cache entry statistics and secondary cache, so it ended up compatible with neither. (`DeleteEntryCached` should not be used here. See how `block_cache->Insert()` is called in block_based_table_reader.cc)

This is an example of drift that can happen with code duplication, in this case from BlockBasedTable::PutDataBlockToCache, rather than trying to share common code. (`cached_block` could be made optional)

Reply:
Thanks for the input. I will create the separate PR for BlockBasedTableBuilder to reuse the code of BlockBasedTableReader.